x-pack/plugins/infra/server/services/log_entries/log_entry_search_strategy.ts

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import * as rt from 'io-ts';
+import { concat, defer, of } from 'rxjs';
+import { concatMap, filter, map, shareReplay, take } from 'rxjs/operators';
+import type {
+  IEsSearchRequest,
+  IKibanaSearchRequest,
+  IKibanaSearchResponse,
+} from '../../../../../../src/plugins/data/common';
+import type {
+  ISearchStrategy,
+  PluginStart as DataPluginStart,
+} from '../../../../../../src/plugins/data/server';
+import { getLogEntryCursorFromHit } from '../../../common/log_entry';
+import { decodeOrThrow } from '../../../common/runtime_types';
+import {
+  LogEntrySearchRequestParams,
+  logEntrySearchRequestParamsRT,
+  LogEntrySearchResponsePayload,
+  logEntrySearchResponsePayloadRT,
+} from '../../../common/search_strategies/log_entries/log_entry';
+import type { IInfraSources } from '../../lib/sources';
+import {
+  createAsyncRequestRTs,
+  createErrorFromShardFailure,
+  jsonFromBase64StringRT,
+} from '../../utils/typed_search_strategy';
+import { createGetLogEntryQuery, getLogEntryResponseRT, LogEntryHit } from './queries/log_entry';
+
+type LogEntrySearchRequest = IKibanaSearchRequest<LogEntrySearchRequestParams>;
+type LogEntrySearchResponse = IKibanaSearchResponse<LogEntrySearchResponsePayload>;
+
+export const logEntrySearchStrategyProvider = ({
+  data,
+  sources,
+}: {
+  data: DataPluginStart;
+  sources: IInfraSources;
+}): ISearchStrategy<LogEntrySearchRequest, LogEntrySearchResponse> => {
+  const esSearchStrategy = data.search.getSearchStrategy('ese');
+
+  return {
+    search: (rawRequest, options, dependencies) =>
+      defer(() => {
+        const request = decodeOrThrow(asyncRequestRT)(rawRequest);
+
+        const sourceConfiguration$ = defer(() =>
+          sources.getSourceConfiguration(dependencies.savedObjectsClient, request.params.sourceId)
+        ).pipe(shareReplay(1));
+
+        const recoveredRequest$ = of(request).pipe(
+          filter(asyncGetRequestRT.is),
+          map(({ id: { esRequestId } }) => ({ id: esRequestId }))
+        );
+
+        const initialRequest$ = of(request).pipe(
+          filter(asyncSubmitRequestRT.is),
+          concatMap(({ params }) =>
+            sourceConfiguration$.pipe(
+              map(
+                ({ configuration }): IEsSearchRequest => ({
+                  params: createGetLogEntryQuery(
+                    configuration.logAlias,
+                    params.logEntryId,
+                    configuration.fields.timestamp,
+                    configuration.fields.tiebreaker
+                  ),
+                })
+              )
+            )
+          )
+        );
+
+        return concat(recoveredRequest$, initialRequest$).pipe(
```

Comment:
:information_source: Only one of them is expected emit anything, so this represents a branch in the observable.